File: app/src/main/java/org/kikermo/thingsaudioreceiver/service/PlayerService.java
package org.kikermo.thingsaudioreceiver.service;

import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.os.IBinder;
import android.support.annotation.Nullable;

import org.kikermo.thingsaudioreceiver.model.data.PlayPosition;
import org.kikermo.thingsaudioreceiver.model.data.Track;
import org.kikermo.thingsaudioreceiver.util.Constants;
import org.kikermo.thingsaudioreceiver.util.Log;
import org.kikermo.thingsaudioreceiver.util.Utils;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;

public class PlayerService extends Service implements MediaPlayer.OnCompletionListener {
    private MediaPlayer mediaPlayer;
    private List<Track> trackList;
    private int trackPointer;
    private boolean repeatList;
    private boolean repeatSong;
    private Disposable progressDisposable;

    public PlayerService() {
        trackList = new ArrayList<>();
        trackPointer = 0;
        repeatList = false;
        repeatSong = false;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        mediaPlayer = new MediaPlayer();
        mediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
        mediaPlayer.setOnCompletionListener(this);


        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addAction(Constants.BA_NEW_TRACKLIST);
        intentFilter.addAction(Constants.BA_SKIP_NEXT);
        intentFilter.addAction(Constants.BA_SKIP_PREV);
        intentFilter.addAction(Constants.BA_PAUSE);
        intentFilter.addAction(Constants.BA_PLAY);
        registerReceiver(controlReceiver, intentFilter);

        progressDisposable = Observable.interval(1, TimeUnit.SECONDS).map(aLong -> mediaPlayer.getCurrentPosition() / 1000)
                .distinctUntilChanged()
                .subscribe(integer -> {
                    PlayPosition playPosition = new PlayPosition();
                    playPosition.setPosition(integer);
                    Intent posInt = new Intent();
                    posInt.putExtra(Constants.BK_PLAYBACKEVENT, playPosition);
                    posInt.setAction(Constants.BA_PLAYBACKEVENT);

                    sendBroadcast(posInt);
                    Log.d("Player", "pos" + integer);
                });
    }


    @Override
    public void onDestroy() {
        mediaPlayer.release();
        mediaPlayer = null;
        unregisterReceiver(controlReceiver);
        if (Utils.notNull(progressDisposable))
            progressDisposable.dispose();
        super.onDestroy();
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }


    @Override
    public void onCompletion(MediaPlayer mediaPlayer) {
        if (repeatSong) {
            mediaPlayer.start();
            return;
        }

        if (trackPointer < trackList.size()) {
            trackPointer++;
            playCurrentSong();
        } else {
            trackPointer = 0;
            if (repeatList)
                playCurrentSong();
        }
    }

    private void playCurrentSong() {
        Track track = trackList.get(trackPointer);

        Intent intent = new Intent();
        intent.putExtra(Constants.BK_PLAYBACKEVENT, track);
        intent.setAction(Constants.BA_PLAYBACKEVENT);

        sendBroadcast(intent);
        try {
            mediaPlayer.reset();
            mediaPlayer.setDataSource(track.getUrl());
            mediaPlayer.prepare();
            mediaPlayer.start();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }


    private BroadcastReceiver controlReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            Log.d("Player", "Command Received");
            switch (intent.getAction()) {
                case Constants.BA_PAUSE:
                    mediaPlayer.pause();
                    break;
                case Constants.BA_PLAY:
                    mediaPlayer.start();
                    break;
                case Constants.BA_SKIP_NEXT:
                    if (trackPointer < trackList.size()) {
                        trackPointer++;
                        playCurrentSong();
                    }
                    break;
                case Constants.BA_SKIP_PREV:
                    if (trackPointer >= 1) {
                        trackPointer--;
                        playCurrentSong();
                    }
                    break;
                case Constants.BA_NEW_TRACKLIST:
                    Log.d("Player", "New list");
                    List<Track> newTracks = intent.getParcelableArrayListExtra(Constants.BK_TRACKLIST);
                    mediaPlayer.stop();
                    trackList.clear();
                    trackList.addAll(newTracks);
                    trackPointer = 0;
                    playCurrentSong();
                    break;
            }
        }
    };
}

File: app/src/main/java/org/kikermo/thingsaudioreceiver/service/PlayerService.kt
package org.kikermo.thingsaudioreceiver.service

import android.app.Service
import android.content.Intent
import android.media.AudioManager
import android.media.MediaPlayer
import android.os.IBinder
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.BehaviorSubject
import org.kikermo.thingsaudio.core.model.Track
import org.kikermo.thingsaudioreceiver.ThingsReceiverApplication
import org.kikermo.thingsaudioreceiver.model.PlayerControlActions
import org.kikermo.thingsaudioreceiver.util.Constants
import timber.log.Timber
import java.io.IOException
import java.util.concurrent.TimeUnit

class PlayerService : Service(), MediaPlayer.OnCompletionListener {
    lateinit var mediaPlayer: MediaPlayer
    private val trackList: MutableList<Track> = mutableListOf()
    private var trackPointer: Int = 0
    private val repeatList: Boolean
    private val repeatSong: Boolean

    lateinit var trackSubject: BehaviorSubject<Track>

    private val compositeDisposable = CompositeDisposable()

    private fun processAction(playerControlActions: PlayerControlActions) {
        Timber.d("Action Received")
        when (playerControlActions) {
            is PlayerControlActions.Pause -> mediaPlayer.pause()
            is PlayerControlActions.Play -> mediaPlayer.start()
            is PlayerControlActions.SkipNext -> if (trackPointer < trackList.size) {
                trackPointer++
                playCurrentSong()
            }
            is PlayerControlActions.SkipPrevious -> if (trackPointer >= 1) {
                trackPointer--
                playCurrentSong()
            }
            is PlayerControlActions.AddSong -> {
                Timber.d("New song")
                trackList.add(playerControlActions.track)
                if (!mediaPlayer.isPlaying) {
                    playCurrentSong()
                }
            }
            is PlayerControlActions.AddTrackList -> {
                Timber.d("New list")
                mediaPlayer.stop()
                trackList.clear()
                trackList.addAll(playerControlActions.trackList)
                trackPointer = 0
                playCurrentSong()
            }
        }
    }

    init {
        trackPointer = 0
        repeatList = false
        repeatSong = false
    }

    override fun onCreate() {
        super.onCreate()
        mediaPlayer = MediaPlayer()
        mediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC)
        mediaPlayer.setOnCompletionListener(this)

        val app = (application as ThingsReceiverApplication) //TODO add dagger

        val playerControlObservable = app.playerControlsObservable
        compositeDisposable.add(playerControlObservable.subscribe(this::processAction, Timber::e))

        trackSubject = app.trackSubject

        compositeDisposable.add(Observable.interval(1, TimeUnit.SECONDS)
            .map { mediaPlayer.currentPosition / 1000 }
            .distinctUntilChanged()
            .subscribe({ integer ->
                val playPosition = integer!!
                val posInt = Intent()
                posInt.putExtra(Constants.BK_PLAYBACKEVENT, playPosition)
                posInt.action = Constants.BA_PLAYBACKEVENT

                sendBroadcast(posInt)
                Timber.d("pos$integer")
            }, Timber::e))
    }

    override fun onDestroy() {
        mediaPlayer.release()

        compositeDisposable.clear()
        super.onDestroy()
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onCompletion(mediaPlayer: MediaPlayer) {
        if (repeatSong) {
            mediaPlayer.start()
            return
        }

        if (trackPointer < trackList.size) {
            trackPointer++
            playCurrentSong()
        } else {
            trackPointer = 0
            if (repeatList)
                playCurrentSong()
        }
    }

    private fun playCurrentSong() {
        val track = trackList[trackPointer]

        trackSubject.onNext(track)

        try {
            mediaPlayer.reset()
            mediaPlayer.setDataSource(track.url)
            mediaPlayer.prepare()
            mediaPlayer.start()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}
